Whitelist std and var for use with custom rolling windows

AlexKirko opened the pull request:
- [X] xref #32865
- [X] 0 tests added / 0 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry
 ### The problem
While researching what funcitons are broken for #32865 , I added `std` and `var` to the list since their output didn't match numpy output. I have since discovered that this is because we default to the sample variance formula for all window calculations.
After closer examination, the algorithm itself turned out to be very robust against custom indexers. It is even resilient against non-monothonic window starts and ends.
There is nothing to do there, so we should revert blacklisting `std` and `var`. I don't think tests are necessary, since we aren't changing anything.
### Food for thought
Some background information to make our decision here more informed:

The reason I first believed these functions to be broken is because using the sample variance formula for sliding windows makes no sense to me from a statistical viewpoint. We use sample variance when the dataset is a sample drawn from a larger population. **A window is not a sample**. When we calculate sliding window variance, we aren't interested in getting the correct variance for some underlying general window, we are interested in computing it correctly for each window, and thus each window is the population.

However, as we discussed with @mroeschke:
1) Usually users are interested in calculating variance for large windows, and the difference between formulas for variance is proportional to `1 / (window_size - 1) - 1 / window_size`
2) We use sample variance as a default everywhere else in pandas.
3) The user can specify `rolling.var(ddof=0)` to set degrees of freedom to zero and get population variance, if they know what they want and are aware that pandas uses sample variance by default.

So the default doesn't make much sense, but it is consistent with the rest of our software, and the harm is negligible for most use cases. The harm in changing the default would be that people who know the package well might expect the sample variance default.

Apologies for the long read. It's a part of my job to find mistakes in data science models, so I'm sensitive to stuff like this.

AlexKirko committed:
stop throwing NotImplemented on std and var

AlexKirko committed:
DOC: edit whatsnew

AlexKirko commented:
/azp run

AlexKirko commented:
Well, better restart the old-fashioned way then...

AlexKirko committed:
restart checks

AlexKirko committed:
restart checks

mroeschke commented:
Could you expand on this test (or create a new one) that checks the result of the `std`/`var` operation? 

https://github.com/pandas-dev/pandas/pull/33180/files#diff-c0ea38b081d5b0c3cf20dc8646f38cefR106

jreback commented:
also if you would like to expand the docs of std/var (for windows) would be helpful (e.g. adding explanations like above)

AlexKirko committed:
TST: add kwargs to tests

AlexKirko committed:
TST: add tests for std and var

AlexKirko commented:
@mroeschke 
Sure! Restructured the test a bit to accomodate the required flexibility.

AlexKirko committed:
DOC: expand documentation on sample variance

AlexKirko commented:
@jreback 
Done. In addition, added a xref to the new note to the expanding windows section, as the user should also be conscious of what they are doing, same as with rolling windows.
Also cleaned up a bit. Bessel-adjusted sample standard deviation is a thautology. All sample estimates of variance are Bessel-adjusted for degrees of freedom.

AlexKirko committed:
CLN: remove trailing whitespace

mroeschke commented on doc/source/user_guide/computation.rst:
```diff
@@ -312,15 +312,35 @@ We provide a number of common statistical functions:
+   Please note that :meth:`~Rolling.std` and :meth:`~Rolling.var` use the sample
+   variance formula by default, i.e. the sum of squared differences is divided by
+   ``window_size - 1`` and not by ``window_size`` during averaging. In statistics,
+   we use sample  when the dataset is drawn from a larger population that we
```
Nit: Looks like you have 2 spaces between `sample` and `when`

AlexKirko replied:
Done.

mroeschke commented on pandas/tests/window/test_base_indexer.py:
```diff
@@ -97,13 +97,57 @@ def get_window_bounds(self, num_values, min_periods, center, closed):
 @pytest.mark.parametrize("constructor", [Series, DataFrame])
 @pytest.mark.parametrize(
-    "func,alt_func,expected",
+    "func,np_func,expected,pd_kwargs,np_kwargs",
```
We can remove `pd_kwargs` for now since it empty in all these cases?

AlexKirko replied:
Done.

AlexKirko committed:
CLN: remove double space

AlexKirko committed:
CLN: remove pd_kwargs from the test

AlexKirko commented:
Numpy dev build is failing. It's being handled in #33507 .

AlexKirko committed:
Merge branch 'master' into rolling-std-var

AlexKirko commented:
@mroeschke @jreback 
All commentds addressed, all green.

AlexKirko commented:
Okay, moving on to the next function.

@jreback, can we merge this PR?

jreback commented:
thanks!
